pkg/controllerhelpers/handlers.go

```diff
@@ -202,3 +203,36 @@ func (h *Handlers[T]) HandleDeleteWithDepth(depth int, obj any, enqueueFunc Enqu
 func (h *Handlers[T]) HandleDelete(obj any, enqueueFunc EnqueueFuncType) {
 	h.HandleDeleteWithDepth(2, obj, enqueueFunc)
 }
+
+// DeletedStronglyTypedFinalStateUnknown is like DeletedFinalStateUnknown, but
+// uses a types.NamespacedName as the key. This is useful for objects that are
+// namespaced, as it allows for more precise identification of the object.
+type DeletedStronglyTypedFinalStateUnknown struct {
+	Key types.NamespacedName
+	Obj interface{}
+}
+
+// DeletionHandlingTypedKeyFunc checks for
+// DeletedStronglyTypedFinalStateUnknown objects before calling
+// MetaNamespaceKeyFuncWithStrongType. This function is essentially the
+// same as the function DeletionHandlingMetaNamespaceKeyFunc
+// but it handles keys of the types.NamespacedName type.
+// This is required for queues that use strongly typed keys of the same type
+func DeletionHandlingTypedKeyFunc(obj interface{}) (types.NamespacedName, error) {
+	if d, ok := obj.(DeletedStronglyTypedFinalStateUnknown); ok {
+		return d.Key, nil
+	}
+	return MetaNamespaceKeyFuncWithStrongType(obj)
+}
+
+// MetaNamespaceKeyFuncWithStrongType is a KeyFunc that returns
+// types.NamespacedName. This function is similar to
+// MetaNamespaceKeyFunc, but it returns a structured key instead of a string.
+// It is useful for queues that use strongly typed keys of the same type.
+func MetaNamespaceKeyFuncWithStrongType(obj interface{}) (types.NamespacedName, error) {
+	objName, err := cache.ObjectToName(obj)
+	if err != nil {
+		return types.NamespacedName{}, err
+	}
+	return types.NamespacedName{Namespace: objName.Namespace, Name: objName.Name}, nil
+}
```

Comment:
seems a bit redundant to use `cache.ObjectToName`
```suggestion
func ObjectToNamespacedName(obj interface{}) (types.NamespacedName, error) {
	metaAccessor, err := meta.Accessor(obj)
	if err != nil {
		return types.NamespacedName{}, fmt.Errorf("object has no meta: %w", err)
	}

	return types.NamespacedName{
		Namespace: metaAccessor.GetNamespace(),
		Name:      metaAccessor.GetName(),
	}, nil
}
```

Reply:
>  This way if the underlying implementation of ObjectToName changes, we would not have to change anything in the code.

I'm not sure I follow. We're in control of the keys in the queue and how we prepare a key from an object implementing metav1.Object. I don't see how a change in the implementation of `cache.ObjectToName` could require a change on our side, can you give me a pointer?

Reply:
This is just over complication on my part but the thought process is this:

So if I am not wrong the proposed changes are basically a 1:1 of the `cache.ObjectToName` function but now it returns a `types.NamespacedName` object. So if there's some change in the `cache.ObjectToName` function(for example if the behaviour of `meta.Accessor` changes), since we've essentially lifted that logic and tweaked it a little to suit our need, wouldn't we also need to make a similar change to our custom implementation? Basically, if what we have copied has changed, we need to change what we have copied.

Again, just an over complication on my end. I'll make the proposed changes if ok with you.

Reply:
Changed the method as per proposal.